homeassistant/components/weather/__init__.py

```diff
@@ -173,46 +419,108 @@ def ozone(self) -> float | None:
         """Return the native unit of measurement for accumulated precipitation."""
-        return self._attr_precipitation_unit
+        return self._attr_native_precipitation_unit
+
+    @property
+    def precipitation_unit(self) -> str | None:
+        """Return the converted unit of measurement for precipitation."""
+        if (precipitation_unit := self._attr_precipitation_unit) is not None:
+            return precipitation_unit
+
+        if (native_precipitation_unit := self.native_precipitation_unit) is None:
+            return None
+
+        if _weather_option_precipitation_uom := self._weather_option_precipitation_uom:
+            return _weather_option_precipitation_uom
+
+        return native_precipitation_unit
 
     @property
     def precision(self) -> float:
         """Return the precision of the temperature value, after unit conversion."""
         if hasattr(self, "_attr_precision"):
             return self._attr_precision
+        if self._override:
+            return (
+                PRECISION_TENTHS
+                if self.hass.config.units.temperature_unit == TEMP_CELSIUS
+                else PRECISION_WHOLE
+            )
         return (
             PRECISION_TENTHS
-            if self.hass.config.units.temperature_unit == TEMP_CELSIUS
+            if self.temperature_unit == TEMP_CELSIUS
             else PRECISION_WHOLE
         )
 
     @final
     @property
-    def state_attributes(self):
+    def state_attributes(self):  # noqa: C901
         """Return the state attributes, converted from native units to user-configured units."""
         data = {}
-        if self.temperature is not None:
+
+        precision = self.precision
+
+        if (temperature := self.temperature) is not None and (
+            temp_unit := self.temperature_unit
+        ) is not None:
             data[ATTR_WEATHER_TEMPERATURE] = show_temp(
                 self.hass,
-                self.temperature,
-                self.temperature_unit,
-                self.precision,
+                temperature,
+                temp_unit,
+                precision,
+            )
+        elif (
+            (temperature := self.native_temperature) is not None
+            and (native_temp_unit := self.native_temperature_unit) is not None
+            and (temp_unit := self.temperature_unit) is not None
+        ):
+            with suppress(ValueError):
+                temperature_f = float(temperature)
+                value_temp = UNIT_CONVERSIONS[CONF_TEMPERATURE_UOM](
+                    temperature_f, native_temp_unit, temp_unit
+                )
+                data[ATTR_WEATHER_TEMPERATURE] = round_temperature(
+                    value_temp, precision
+                )
+
+        if (temp_unit := self.temperature_unit) is not None:
+            data[ATTR_WEATHER_TEMPERATURE_UNIT] = (
+                self.hass.config.units.temperature_unit if self._override else temp_unit
```

Comment:
Why do we check `self._override` here, but not for pressure, wind_speed, etc.?

Reply:
In legacy it has the config temp_unit and it provided conversion.
In new setup it will only convert if the user ask's for it (via frontend).

In legacy for the other uom's there was no conversion so it's enough to just display them.

Reply:
Logic is reversed and changed so no need for override anymore